### PR TITLE
 Resolve gate-sync-online issues

### LINKED ISSUE #44: Story: Sync karma gates between players

### PR DESCRIPTION
**Pull request by 6fears7**
closes #44

Requires testing

### CODE REVIEW

File: Game/RainMeadow.GameplayHooks.cs [hunk 1]
```diff
@@ -15,6 +15,36 @@ public void GameplayHooks()
             On.Creature.Grasp.ctor += GraspOnctor;
             On.PhysicalObject.Grabbed += PhysicalObjectOnGrabbed;
             On.Creature.SuckedIntoShortCut += CreatureSuckedIntoShortCut;
+            On.RegionGate.Update += RegionGateTransition;
+        }
+
+        private void RegionGateTransition(On.RegionGate.orig_Update orig,  RegionGate self, bool eu)
```
Comment (Chrometurtle) [inline Game/RainMeadow.GameplayHooks.cs:21]: Throw this into RainMeadow/Story/RainMeadow.StoryHooks.cs

I've been keeping Story only gamehooks here

**Comment by 6fears7** [comment]
https://github.com/henpemaz/Rain-Meadow/blob/gate-sync-online/Online/Resource/Lobby.cs#L14 

TODO: Check with @henpemaz

File: Story/RainMeadow.StoryHooks.cs [hunk 2]
```diff
@@ -171,9 +180,71 @@ private bool AllPlayersThroughtoOtherSide(On.RegionGate.orig_AllPlayersThroughTo
...
+                orig(self);
+                return -1;
+
+            }
+
+            int num = -1;
+            foreach (var playerAvatar in OnlineManager.lobby.playerAvatars)
+            {
+
+                int num2 = self.DetectZone(playerAvatar.Value); // wants an AbstractCreature
```
Comment (6fears7) [inline Story/RainMeadow.StoryHooks.cs:201]: @Chrometurtle, DetectZone wants a parameter of type AbstractCreature. Even after an attempted cast, the compiler warns that playerAvatar.Value is of type OnlineCreature. Wanted to pick your brain on that.

File: Story/RainMeadow.StoryHooks.cs [hunk 3]
```diff
@@ -245,8 +312,28 @@ private bool PlayersStandingStill(On.RegionGate.orig_PlayersStandingStill orig,
             return true;
         }
 
+
+
+
+        private static void HUDILKarmaBlinkRed(ILContext il) // TODO: Build out HUD and rewrite this
```
Comment (6fears7) [inline Story/RainMeadow.StoryHooks.cs:318]: @Chrometurtle for custom multiplayer HUD, where do you want to place that file?

**Review by 6fears7 — COMMENTED** [review]
Please review at your convenience, my liege.

File: Story/RainMeadow.StoryHooks.cs (hunk 2 above)
Comment (Chrometurtle) [reply Story/RainMeadow.StoryHooks.cs:201]: playerAvatar.Value.realizedCreature.abstractCreature

Just keeping records for people to see if they're interested

File: Menu/LobbySelectMenu.cs [hunk 4]
```diff
@@ -213,6 +213,12 @@ public override void Update()
 
         private void Play(SimplerButton obj)
         {
+            if (ModManager.JollyCoop)
+            {
+                ShowErrorDialog("Please disable JollyCoop before playing Online");
```
Comment (6fears7) [inline Menu/LobbySelectMenu.cs:218]: @Chrometurtle I added a check to ensure that Jolly was disabled before one could press Play to protect against unexpected behavior. Thoughts? I can push in separate PR if we want to keep RegionGate activity clean.

Comment (Chrometurtle) [reply Menu/LobbySelectMenu.cs:218]: @henpemaz :D?

**Comment by 6fears7** [comment]
depends on #53

File: Story/RainMeadow.StoryHooks.cs (hunk 2 above)
Comment (henpemaz) [reply Story/RainMeadow.StoryHooks.cs]: can probably be simplified to `(playerAvater.Value.apo as AbstractCreature)`

**Comment by henpemaz** [comment]
As there's a TON of general fixes in this branch we're merging it without fully closing #44 and #45

**Pull request merged**